core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -259,6 +261,209 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+
+    Map<Double, SEXP> methods = findFunctionMatchingAnyOfSignatures(context, functionEnv, allPossibleSignatures, 0, null);
+
+    if(methods.isEmpty()) {
+      throw new EvalException("object of type 'S4' is not subsettable");
+    }
+
+    TreeMap<Double, SEXP> map = new TreeMap<>(methods);
+    function = map.get(map.firstKey());
+
+    allArgs.add(source);
+    allArgs.add(args.getElementAsSEXP(1));
+    SEXP result = context.evaluate(new FunctionCall(function, allArgs.build()));
+    return result;
+  }
+
+  private static Map<Double, SEXP> findFunctionMatchingAnyOfSignatures(Context context, Environment functionEnv, Object[][] signatures, int useSignature, Map<Double, SEXP> methods) {
+    // recursively go through arguments and get their class and lookup until the signature is found.
+    // Issues:
+    // - Stops with the first hit
+    // - Is not aware of distance between classes of input arguments to different method signatures
+    if(methods == null) {
+      methods = new HashMap<>(signatures.length);
+    }
+    if(useSignature < signatures.length) {
+      Object[] currentSig = signatures[useSignature];
+      int currentSigLength = currentSig.length;
+      Symbol methodName = Symbol.get((String) currentSig[0]);
+      SEXP function = functionEnv.findVariable(context, methodName);
+
+      if(function != Symbol.UNBOUND_VALUE) {
+        int[] distances = new int[currentSigLength - 1];
+        for(int i = 0; i < currentSigLength - 1; i++) {
+          distances[i] = (Integer) currentSig[i + 1];
+        }
+
+        double methodRating = computeMethodRank(distances);
+        methods.put(methodRating, function);
+      }
+      useSignature++;
+      return findFunctionMatchingAnyOfSignatures(context, functionEnv, signatures, useSignature, methods);
+    }
+    return methods;
+  }
+
+  private static double computeMethodRank(int[] distances) {
+    double methodRank = 0.0;
+    boolean hasZero = false;
+
+    for(int i = 0; i < distances.length && hasZero == false; i++) {
+      if (distances[i] == 0) {
+        hasZero = true;
+      }
+    }
+
+    int totalDist = 0;
+    for(int i = 0; i < distances.length; i++) {
+      totalDist = totalDist + distances[i];
+    }
+    methodRank = methodRank + ((double)totalDist * 1000.0);
+
+    double initRate = 10.0;
+    for(int i = 0; i < distances.length; i++) {
+      methodRank = methodRank + (initRate / ((double)i + 1)) * (double)distances[i];
+    }
+
+    if(!hasZero) {
+      methodRank = methodRank + 10000.00;
+    }
+
+    return methodRank;
+  }
+
+  public static Object[][] generateAllPossibleSignatures(Context context, Environment rho, PairList args,
+                                                         Object[][] signature, int depth, int current) {
+    if(current < depth) {
+      String evaluatedArg = evaluateAndGetClass(context, args, rho, current);
```

Comment:
It looks like an argument could be evaluated more than once. Can you add some tests with arguments that have side effects? If dispatching requires evaluating the arguments, then why not just evaluate all arguments up front?